### ISSUE TITLE
Docker version 17.06.1 or later required - Docker location is in path

### ISSUE THREAD

**Issue by rhartman-qwinix** [microsoft/vscode-remote-release#3941]
- VSCode Version: 1.50.1
- Local OS Version: Pop! OS 20.10
- Remote OS Version: Alpine 3.12
- Remote Extension/Connection Type: Docker

```
Version: 1.50.1
Commit: d2e414d9e4239a252d1ab117bd7067f125afd80a
Date: 2020-10-13T14:44:48.716Z
Electron: 9.2.1
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Linux x64 5.8.0-7625-generic
```

Steps to Reproduce:

1. select: Remote-Containers: Attach to a running container
Error shows: 'Docker version 17.06.1 or later required'

Does this issue occur when you try this locally?: Yes
Does this issue occur when you try this locally and all extensions are disabled?: Yes

```
 echo $PATH
/var/lib:/home/devuser/.pyenv/plugins/pyenv-virtualenv/shims:/home/devuser/.pyenv/shims:/home/devuser/.pyenv/bin:/home/devuser/.local/bin:/var/lib:/home/devuser/.pyenv/plugins/pyenv-virtualenv/shims:/home/devuser/.pyenv/shims:/home/devuser/.pyenv/bin:/home/devuser/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```
Pretty much any of the "Remote-Containers" commands ends in this error. It seems that VS code does not recognize my docker install

I have seen several similar issues, but they were all on mac. As you can see the docker executable is in my /var/lib directory

```
devuser@devuser-pop-os:~/Projects/removed/cxEngine$ ls -la /var/lib/
total 280
<SNIP>
drwx--x--x 15 root          root          4096 Nov  3 10:21 docker
<SNIP>
```

From terminal in vscode:
```
devuser@devuser-pop-os:~/Projects/removed/cxEngine$ docker -v
Docker version 19.03.13+azure, build cd8016b6bcc51a51f577bcffb37e6a870f7813f9
```

whereis docker?
```
devuser@devuser-pop-os:~/Projects/removed/cxEngine$ whereis docker
docker: /usr/bin/docker /etc/docker /usr/libexec/docker /var/lib/docker /usr/share/man/man1/docker.1.gz
```

EDIT:
Note: if someone could explain how I can get more logs out of vsCode so I can track this error, that would be much appreciated. I could track this down a little easier if I had a log to see what triggered that error.

**Comment by chrmarti** [comment]
How are you launching VS Code? From a shell or from the UI? Does it work with any of the two?

Also check where you are adding `/var/lib` to the PATH, it seems VS Code gets a PATH that does not have that.

**Comment by rhartman-qwinix** [comment]
@chrmarti I am typically launching VS Code from the UI, but every so often it will be from the CLI. 

I retested the scenario by launching from the CLI and I got the same results.

I am adding `/var/lib` to the PATH inside of the `.bashrc` file. I agree that it does not seem to recognize this path, any suggestions on how I can debug what it is reading?

Depending on how you want to define where the docker executable and where it is on the machine, I think I have all the bases covered in the PATH

```
devuser@devuser-pop-os:~$ whereis docker
docker: /usr/bin/docker /etc/docker /usr/libexec/docker /usr/share/man/man1/docker.1.gz
devuser@devuser-pop-os:~$ which docker
/usr/bin/docker
devuser@devuser-pop-os:~$ echo $PATH
/usr/bin/docker:/home/devuser/.pyenv/plugins/pyenv-virtualenv/shims:/home/devuser/.pyenv/shims:/home/devuser/.pyenv/bin:/home/devuser/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```

**Comment by wherka** [comment]
Hello @chrmarti,
from what I can see the error reporter is bumping into the very same problem as myself. I believe it has nothing to do with the client i.e. docker. This is most likely about the daemon. At least that's what's bugging me.

Here is the symptom:
![image](https://user-images.githubusercontent.com/1831764/98416297-b8960000-2076-11eb-8ef9-0d92f23f4392.png)


This is what's on the server side in my case:

>docker info
**Server Version: 1.13.1**

>lsb_release -a
LSB Version:    :core-4.1-amd64:core-4.1-noarch
Distributor ID: **RedHatEnterpriseServer**
Description:    Red Hat Enterprise Linux Server release 7.7 (Maipo)
Release:        **7.7**
Codename:       Maipo

>oc version
**oc v3.11.154**
kubernetes v1.11.0+d4cacc0

This docker daemon installation comes with or is implied by OpenShift 3.x on RHEL 7.7, so that's not something I can change easily without nasty consequences on the OC side.

I know exactly what is wrong there and I know how to circumvent this to make it talk to even this old daemon. However, according to the license agreement I'm legally not allowed to do it as that's closed source:

https://ms-vscode-remote.gallerycdn.vsassets.io/extensions/ms-vscode-remote/remote-containers/0.148.0/1604006153365/Microsoft.VisualStudio.Services.Content.License

7.	SCOPE OF LICENSE. (...)
You may not:
    ***	work around any technical limitations in the software;**

Any chance of lifting this version check and popping up just a gentle warning, so the users who can upgrade the docker daemon can do the needful and these who cannot as myself can carry on working in a best effort mode please?

Barring the arbitrary versions of Docker is just not nice to us poor end users who cannot benefit from such a great extension otherwise. It is really sad :( 
I really love working with remote-containers extension and it just breaks my heart seeing this popup - especially when I know most of the features and in fact all the features I care about will work just fine if you let me go past this version check.

Is there something you could perhaps consider please? I would really appreciate it.

Kind Regards,
Waldek

**Comment by PavelSosin-320** [comment]
The problem is not in the Docker implementation version but in **Docker specification level**. The current Docker specification version is 19.03 and it is 1.Docker capabilities, 2.Docker engine API specification version and 3.docker client version.  Reading Docker documentation you can see frequently "This otption is supported only if both Engine and client are above 19... spec."
Upgrading of Docker version to the recently released spec is not optional. On all platforms upgrading of Docker is well documented and stable process. Upgrading to the latest released version is not harmfull. I'm sure that if you familiar with Docker features you can upgrade your Docker version easily. It is pre-condition of normal customer support - nobody cab afford to support more than few specification back except cases of exceptionally strict corporate SLAs.

**Comment by wherka** [comment]
@PavelSosin-320 : thanks for your comment. This is exactly what I'm referring to - "strict corporate SLAs".  Here this is not necessarily SLAs, but corporate restrictions. I do come from corporate world and that's the problem.

As far as upgrading the Docker and in fact any other packages, I don't agree with you. Pretty much any enterprise level Linux distributions(in my case RHEL) can only pull the packages from the official channels. The moment you start interfering with this process, you are on your own as far as the support. Platform/OS providers would only support what they are responsible for in terms of packaging. This is not about the technical ability to upgrade the software - I can do it in couple of minutes, but by doing so I'm voiding the license agreements, which is a bit risky in the corporate world.

On the other hand - I'm 100% sure that the extension works just fine when I send the higher version number. I tested it by recompiling the client and touching nothing on the server side, so it's not just the assumption. I know it works.

I understand your position in terms of supporting the just a sensible subset of versions too. What I don't understand is the lack of ability to override this check. We are not really in the license agreement and there is no SLA. If the extension is not working as it should because the version is too low and I've been warned. Warning is fine, the current popup and bailing out is not great.

Anyway, thanks for your input. I appreciate it.

**Comment by rhartman-qwinix** [comment]
> The problem is not in the Docker implementation version but in **Docker specification level**. The current Docker specification version is 19.03 and it is 1.Docker capabilities, 2.Docker engine API specification version and 3.docker client version. Reading Docker documentation you can see frequently "This otption is supported only if both Engine and client are above 19... spec."

Can you be more specific on the requirements for the client and the engine?
```
devuser@devuser-pop-os:~$ docker version
Client:
 Version:           19.03.13+azure
 API version:       1.40
 Go version:        go1.13.15
 Git commit:        cd8016b6bcc51a51f577bcffb37e6a870f7813f9
 Built:             Tue Sep 15 23:02:04 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          3.0.13+azure
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.11
  Git commit:       77e06fda0c
  Built:            Mon Jun  1 20:22:00 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Comment by wherka** [comment]
@rhartman-qwinix : you've just nailed it!  The problem here is we seem to be comparing apples to oranges as far as the version..

The actual call(from the docker cli) which is used to get the version in this extension:
 **docker version --format {{.Server.Version}}**

In your case it's going to end up with: **3.0.13+azure**, in mine: **1.13.1**. In either case that's an internal version completely out of line when it comes to the official package released by Mirantis(formerly docker.com).

This begs the question: should we even use this sort of comparison at this stage to determine if vscode remote containers will work there or not? I'd say that's pointless. We should just give the end user a hint that the extension may or may not work or to call it say: "degraded mode", but since the versioning on the Engine part is not really consistent depending on the vendor, that's really misleading to leave it as is.

I'm in a similar situation as you @rhartman-qwinix  and I would indeed benefit from having such a fantastic utility in my toolkit. If I have to, I'll recompile the client to fake this version report, but honestly I will not recommend anybody should follow. It's better to fix it at source.

**Comment by wherka** [comment]
@chrmarti : I've just found an interesting ticket: https://github.community/t/what-really-is-docker-3-0-6/16171
It shows the origin of the **3.0.13+azure** installation. It's a short read but it proves my point about the necessity to make it possible to lift the restriction on the engine version check as it's clearly not reliable.
In this case it's Microsoft who is shipping the Docker the way it breaks the closed-source VSCode extension which is developed and maintained by Microsoft. We made a full circle here.
I really hope we can make a progress here as you will make many people really happy by doing so. In my organisation alone we can easily see hundreds of developers potentially benefitting from this change. I created another ticket which is more explicitly referring to this problem: https://github.com/microsoft/vscode-remote-release/issues/3970

**Comment by PavelSosin-320** [comment]
I understand this is Azure issue. They use very old Docker version in Azure.  If Docker client and Docker engine specification versions are don't match many things will not work. Actually you propose to check Docker API version instead of Docker version. Docker client 19.03 uses API version 1.40 as you can see in the Docker info output.:
PS C:\WINDOWS\system32> docker version
Client:
 Version:           19.03.3
 **API version:       1.40**
Server:
....... 2375/**v1.40**/info:
Theoretically, Docker API server can support several API versions on several service points
If you read the "what really is docker 3-0-06" you saw that it is legal issue :( - Azure can't use recent Docker CE and bill customers.
I suppose if API spec is OK every VSCode issue is a bug, otherwise Azure doesn't sutisfy installation requirements - the error message is correct.

**Comment by wherka** [comment]
@PavelSosin-320 : using the highest supported Engine API as a check is definitely more accurate indeed as here we're talking about comparing the versions from entirely different domains. Docker Engine CE vs internal Azure versions which seem to have absolutely nothing in common. 3.x has nothing to do with 3.x of Docker CE. BTW 3.x of Docker CE was never released as docker.com changed the versioning conventions after 1.13 as far as I remember (1.13 -> 17.x).

When you say "They use very old Docker version in Azure.", I don't think that's correct. 1.40 in Docker Engine API maps the 19.03.5 Docker CE, which I believe was released quite recently(weeks perhaps):
https://docs.docker.com/engine/api/#api-version-matrix

They(Microsoft) seem to explain their releasing strategy in the post mentioned earlier where they say they often build from fresh sources and upgrade it regularly, so I don't think that's actually a problem. The problem is diverging from the version conventions from unknown(to me) reasons. Or if you take another angle - taking the release number of the Engine rather than the highest supported Engine API as a base for comparison in the ms-vscode-remote.remote-containers extension.

Assuming the version check is corrected, the original problem raised by @rhartman-qwinix  would be fixed.

However, I'm not completely happy with this approach since I truly believe we should have an override to give other people who have to work with lower Engine API than supported, like myself a chance to work in a best effort mode by acknowledging a possibly degraded set of features, but exposing them to such a fantastic extension nonetheless.

If you are curious to why I'm so insistent on introducing such an override, please check the following ticket: https://github.com/microsoft/vscode-remote-release/issues/3970

I would really appreciate your comments there too.

**Comment by chrmarti** [comment]
Fixing by checking the API version is at least 1.35.

**Comment by rhartman-qwinix** [comment]
@chrmarti Thank you! Would I have to compile the module to get your update on my machine? Having this working on my dev setup right now would really help me out professionally. Can you give me any tips on what I can do to get it running locally?

Thanks again!

**Comment by chrmarti** [comment]
@rhartman-qwinix The fix will be in the next version (0.150.0) that will at first be available for VS Code Insiders only (later for VS Code 1.52).

**Comment by chrmarti** [comment]
Verification: By code inspection.
